timed/src/lib.rs

```diff
@@ -91,60 +127,99 @@ impl Trace {
 
-    pub fn dump(id: &str, processor: fn(&str), stats: &TracingStats) {
-        let start = std::time::Instant::now();
+    pub fn dump(id: &str, options: &TraceOptions) {
+        if options.chrome_trace.is_none() && options.statistics.is_none() { return; }
+
         let mut traces = TRACES.lock().unwrap();
         let entry = traces.entry(id.to_string()).or_insert(vec![]);
+        let mut stats_map = HashMap::new();
+        let mut total_time_nanos: u128 = 0;
+        let mut chrome_trace_string = String::new();
+
+        if options.chrome_trace.is_some() {
+            chrome_trace_string.push_str("[\n");
+        }
 
-        processor("[");
         for (i, hop) in entry.iter().enumerate() {
-            let is_last = i == entry.len() - 1;
-            let trace = format!(
-                "{{ \"pid\": 0, \"ts\": {},  \"ph\": \"{:?}\", \"name\": \"{}\" }}",
-                hop.ts, hop.ph, hop.name
-            );
-            processor(&format!("    {}{}", trace, if !is_last { "," } else { "" }));
+            if options.statistics.is_some() {
+                if let Finish(d) = hop.ph {
+                    stats_map.entry(hop.name.clone())
+                        .or_insert(vec![])
+                        .push(d);
+                    total_time_nanos += d.as_nanos();
+                }
+            }
+
+            if options.chrome_trace.is_some() {
+                let is_last = i == entry.len() - 1;
+                let trace = format!(
+                    "{{ \"pid\": 0, \"ts\": {},  \"ph\": \"{}\", \"name\": \"{}\" }}",
+                    hop.ts, hop.ph.to_string(), hop.name
+                );
+                chrome_trace_string.push_str(&format!("    {}{}\n", trace, if !is_last { "," } else { "" }));
+            }
         }
-        processor("]");
-        processor(&format!("Dumping traces took {:?}", start.elapsed()));
 
-        match stats {
-            TracingStats::None => {}
-            TracingStats::Percentage => {}
+        if options.chrome_trace.is_some() {
+            chrome_trace_string.push_str("]");
         }
+
+        options.chrome_trace.map(|f| f(&chrome_trace_string));
+        options.statistics.map(|f| Trace::print_statistics(&f, &stats_map, total_time_nanos));
     }
 
-    pub fn new(id: &str, processor: Option<fn(&str)>, stats: Option<TracingStats>) -> Trace {
+    pub fn new(id: &str, options: TraceOptions) -> Trace {
         let trace = Trace {
             id: id.into(),
-            processor: processor.unwrap_or(|x: &str| {
-                println!("{}", x);
-            }),
-            stats: stats.unwrap_or(TracingStats::None),
+            options,
         };
         Self::register(&trace.id);
         trace
     }
-}
 
-impl Drop for Trace {
-    fn drop(&mut self) {
-        Trace::dump(&self.id, self.processor, &self.stats);
+    fn print_statistics(processor: &fn(&str), stats_map: &HashMap<String, Vec<Duration>>, total_time_nanos: u128) {
+        struct FnStats {
```

Comment:
I am not familiar with this idiom. Does this type definition get instantiated at every function invocation? Any reason you do not declare it above without the pub qualifier?

Reply:
It's a nice feature that lets you define a struct (or any type) within a scope. I use it when the struct is only meaningful for the current scope and nowhere else, like here: `FnStats` only exists so I can sort the statistics later on, and nothing else. It;s definition is done at compile time, just like any other struct.

Reply:
It's a nice feature that lets you define a struct (or any type) within a scope. I use it when the struct is only meaningful for the current scope and nowhere else, like here: `FnStats` only exists so I can sort the statistics later on, and nothing else. It;s definition is done at compile time, just like any other struct.